internal/pkg/earpointer/taint.go

```diff
@@ -0,0 +1,167 @@
+// When a struct object contains taint fields, these fields and all their
+// subfields are included in taint sources, but the struct object itself is
+// not a taint source. However, if this object has no fields, then this object
+// is considered as a taint source.
+//
+// For example, consider object of type "struct {x: *T, y *T}" where x is a
+// taint field, and this object's heap is "{t0,t1}: [x:int->t2, y:int->t3],
+// {t2, t4} --> t5", then {t2, t4, t5} are taint sources.
+func GetSrcRefs(src *source.Source, state *Partitions, isTaintField func(named *types.Named, index int) bool,
+	callees map[*ssa.Function]bool) ReferenceSet {
+	refs := make(ReferenceSet)
+	val, ok := src.Node.(ssa.Value)
+	if !ok {
+		return nil
+	}
+	rep := state.Representative(MakeLocalWithEmptyContext(val))
+	var getStructFieldRefs func(ref Reference, tp types.Type)
+	getStructFieldRefs = func(ref Reference, tp types.Type) {
+		switch tp := tp.(type) {
+		case *types.Named:
+			if tt, ok := tp.Underlying().(*types.Struct); ok {
+				// Look for the taint fields.
+				hasTaintField := false
+				for i := 0; i < tt.NumFields(); i++ {
+					f := tt.Field(i)
+					if isTaintField(tp, i) {
+						for fd, fref := range state.PartitionFieldMap(state.Representative(ref)) {
+							if fd.Name == f.Name() {
+								refs[fref] = true
+								// Mark all the subfields to be tainted.
+								getFieldReferences(fref, refs, state, callees, make(ReferenceSet))
+								hasTaintField = true
+							}
+						}
+					}
+				}
+				if !hasTaintField {
+					getFieldReferences(ref, refs, state, callees, make(ReferenceSet))
+				}
+			}
+		case *types.Pointer:
+			if r := state.PartitionFieldMap(rep)[directPointToField]; r != nil {
+				getStructFieldRefs(r, tp.Elem())
+			} else {
+				getStructFieldRefs(rep, tp.Elem())
+			}
+		case *types.Array:
+			for _, r := range state.PartitionFieldMap(rep) {
+				getStructFieldRefs(r, tp.Elem())
+			}
+		case *types.Slice:
+			for _, r := range state.PartitionFieldMap(rep) {
+				getStructFieldRefs(r, tp.Elem())
+			}
+		case *types.Chan:
+			for _, r := range state.PartitionFieldMap(rep) {
+				getStructFieldRefs(r, tp.Elem())
+			}
+		case *types.Map:
+			for _, r := range state.PartitionFieldMap(rep) {
+				getStructFieldRefs(r, tp.Elem())
+			}
+		case *types.Basic, *types.Tuple, *types.Interface, *types.Signature:
+			// These types do not currently represent possible source types
+		}
+	}
+
+	getStructFieldRefs(rep, val.Type())
+	return refs
+}
+
+// Obtains all the field references and their aliases for "ref".
+// For example, return {t0, t5, t1, t3, t4} for "{t0,t5}: [0:int->t1, 1:int->t3], {t3} --> t4".
+func getFieldReferences(ref Reference, result ReferenceSet, state *Partitions, callees map[*ssa.Function]bool, visited ReferenceSet) {
+	visited[ref] = true
+	for _, m := range state.PartitionMembers(ref) {
+		if isWithinFunctions(m, callees) {
+			result[m] = true
+		}
+	}
+	rep := state.Representative(ref)
+	result[rep] = true
```

Comment:
Doesn't `PartitionMembers` include the `rep`? Or do we want to give the `rep` special status such that it does not need to be `withinFunctions`?